<a href="https://colab.research.google.com/github/mspendyala/data-engineering/blob/main/ExploratoryDataAnalysis/DetectAndRemoveOutliers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Detecting and Removing Outliers**

Detect and remove outliers using Z-score and IQR score

In [8]:
# Import needed libraries
import pandas as pd
from sklearn import datasets
from scipy import stats
import numpy as np
import sklearn

In [18]:
# Import california housing price dataset
# x is independent variables are many where as y is a dependent variable which is always one
california = sklearn.datasets.fetch_california_housing()
x = california.data
y = california.target
columns = california.feature_names

In [24]:
# Create the dataframe
california_df = pd.DataFrame(california.data)
california_df.columns = columns
california_df['prediction_target'] = y
california_df.head()

MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  prediction_target  
0    -122.23              4.526  
1    -122.22              3.585  
2    -122.24              3.521  
3    -122.25              3.413  
4    -122.25              3.422

**Using Z-Score**

What is Z-Score? 
How many standard deviations is the given value is this from the mean is called Z-Score. If Z-Score is zero then it is the same as the mean.

In [28]:
# Use Z-Score function defined in sciopy library to detect the outliers
# zscore could be above or below mean we are only taking absolute value

california_df_zscore = california_df
z_score = np.abs(stats.zscore(california_df))
print(z_score)

[[2.34476576 0.98214266 0.62855945 ... 1.05254828 1.32783522 2.12963148]
 [2.33223796 0.60701891 0.32704136 ... 1.04318455 1.32284391 1.31415614]
 [1.7826994  1.85618152 1.15562047 ... 1.03850269 1.33282653 1.25869341]
 ...
 [1.14259331 0.92485123 0.09031802 ... 1.77823747 0.8237132  0.99274649]
 [1.05458292 0.84539315 0.04021111 ... 1.77823747 0.87362627 1.05860847]
 [0.78012947 1.00430931 0.07044252 ... 1.75014627 0.83369581 1.01787803]]


8.3252	is 2.34476576 standard deviations away from mean of the MedINc column